File: cpp/111-120/Populating Next Right Pointers in Each Node.cpp
class Solution {
public:
    void connect(TreeLinkNode *root) {

        TreeLinkNode *ancestor = root;
        while(ancestor)
        {
            TreeLinkNode *tln = ancestor;
            if(!tln->left) break;
            while(tln)
            {
                tln->left->next = tln->right;
                if(tln->next) tln->right->next = tln->next->left;
                tln = tln->next;
            }
            ancestor = ancestor->left;
        }
    }
};

File: python/111-120/Populating Next Right Pointers in Each Node.py
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None

class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if not root:
            return
        
        while root.left:
            cur = root
            pre = None
            while cur:
                if pre:
                    pre.next = cur.left
                cur.left.next = cur.right
                pre = cur.right
                cur = cur.next
            root = root.left